src/server/middleware/blob/put_blob_upload.go

```diff
@@ -15,17 +15,53 @@
 package blob
 
 import (
-	"net/http"
-	"strconv"
-
+	"fmt"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/lib/errors"
 	"github.com/goharbor/harbor/src/lib/log"
+	blob_models "github.com/goharbor/harbor/src/pkg/blob/models"
 	"github.com/goharbor/harbor/src/pkg/distribution"
 	"github.com/goharbor/harbor/src/server/middleware"
+	"github.com/goharbor/harbor/src/server/middleware/requestid"
+	"net/http"
+	"strconv"
 )
 
-// PutBlobUploadMiddleware middleware to create Blob and ProjectBlob after PUT /v2/<name>/blobs/uploads/<session_id> success
+// PutBlobUploadMiddleware middleware to create Blob and ProjectBlob after PUT /v2/<name>/blobs/uploads/<session_id>?digest=<digest> success
 func PutBlobUploadMiddleware() func(http.Handler) http.Handler {
-	return middleware.AfterResponse(func(w http.ResponseWriter, r *http.Request, statusCode int) error {
+
+	before := middleware.BeforeRequest(func(r *http.Request) error {
+		ctx := r.Context()
+		logger := log.G(ctx)
+
+		v := r.URL.Query()
+		digest := v.Get("digest")
+
+		// digest empty is handled by the blob controller GET method
+		bb, err := blob.Ctl.Get(r.Context(), digest)
+		if err != nil {
+			if errors.IsNotFoundErr(err) {
+				return nil
+			}
+			return err
+		}
+
+		switch bb.Status {
+		case blob_models.StatusNone, blob_models.StatusDelete, blob_models.StatusDeleteFailed:
+			err := blob.Ctl.Touch(r.Context(), bb)
```

Comment:
If its status is `None` you don't need to `Touch` it, correct?

Reply:
it can be discussed.